Library/Homebrew/cmd/list.sh

```diff
@@ -0,0 +1,85 @@
+#:  * `list`
+#:
+#: List all installed formulae and casks.
+#: If <formula> is provided, summarise the paths within its current keg.
+#: If <cask> is provided, list its artifacts.
+
+# HOMEBREW_CELLAR, HOMEBREW_PREFIX are set by brew.sh
+# shellcheck disable=SC2154
+
+homebrew-list() {
+  case "$1" in
+    # check we actually have list and not e.g. listsomething
+    list) ;;
+    list*) return 1 ;;
+    *) ;;
+  esac
+
+  local ls_args=()
+  local formula=""
+  local cask=""
+
+  while [[ "$#" -gt 0 ]]
+  do
+    case "$1" in
+      # check for flags passed to ls
+      -1 | -l | -r | -t)
+        ls_args+=("$1")
+        shift
+        ;;
+      --formula)
+        formula=1
+        shift
+        ;;
+      --cask)
+        cask=1
+        shift
+        ;;
+      # reject all other flags
+      *) return 1 ;;
+    esac
+  done
+
+  local tty
+  if [[ -t 1 ]]
+  then
+    tty=1
+  fi
+
+  if [[ -z "${cask}" && -d "${HOMEBREW_CELLAR}" ]]
+  then
+    if [[ -n "${tty}" && -z "${formula}" ]]
+    then
+      ohai "Formulae"
+    fi
+
+    local formula_output
+    formula_output="$(ls "${ls_args[@]}" "${HOMEBREW_CELLAR}")"
+    if [[ -n "${formula_output}" ]]
+    then
+      echo "${formula_output}"
+    fi
+
+    if [[ -n "${tty}" && -z "${formula}" ]]
+    then
+      echo
+    fi
+  fi
+
+  if [[ -z "${formula}" && -d "${HOMEBREW_CASKROOM}" ]]
+  then
+    if [[ -n "${tty}" && -z "${cask}" ]]
+    then
+      ohai "Casks"
+    fi
+
+    local cask_output
+    cask_output="$(ls "${ls_args[@]}" "${HOMEBREW_CASKROOM}")"
+    if [[ -n "${cask_output}" ]]
+    then
+      echo "${cask_output}"
+    fi
+
+    return $?
```

Comment:
Probably want to capture the return code for `ls` and not `echo` here. Any maybe do it also for the `formula` `if` block above? But then there will be two possible return codes to return to `brew.sh` that will need combining somehow.

Reply:
@carlocab Good point, have refactored this with early `return`s to better match how `safe_system` behaves.